File: apps/web/src/lib/date.ts
import type { Subscription } from '@/types'

const shortDateFormatter = new Intl.DateTimeFormat('en', { month: 'short', day: 'numeric' })

export function greetingForTime(date = new Date()) {
  const hour = date.getHours()
  if (hour < 5) return 'Good night'
  if (hour < 12) return 'Good morning'
  if (hour < 17) return 'Good afternoon'
  if (hour < 21) return 'Good evening'
  return 'Good night'
}

export function nextBillingDate(billingDay: number, from = new Date()) {
  const today = new Date(from.getFullYear(), from.getMonth(), from.getDate())
  let result = billingDate(from.getFullYear(), from.getMonth(), billingDay)
  if (result < today) result = billingDate(from.getFullYear(), from.getMonth() + 1, billingDay)
  return result
}

export function subscriptionNextBillingDate(subscription: Pick<Subscription, 'billing_day' | 'next_payment_date'>, from = new Date()) {
  const persisted = subscription.next_payment_date ? parseCalendarDate(subscription.next_payment_date) : null
  const today = new Date(from.getFullYear(), from.getMonth(), from.getDate())
  return persisted && persisted >= today ? persisted : nextBillingDate(subscription.billing_day, from)
}

export function formatShortDate(date: Date) {
  return shortDateFormatter.format(date)
}

function billingDate(year: number, month: number, preferredDay: number) {
  const lastDay = new Date(year, month + 1, 0).getDate()
  return new Date(year, month, Math.max(1, Math.min(preferredDay, lastDay)))
}

function parseCalendarDate(value: string) {
  const match = /^(\d{4})-(\d{2})-(\d{2})/.exec(value)
  if (!match) return null
  const date = new Date(Number(match[1]), Number(match[2]) - 1, Number(match[3]))
  return Number.isNaN(date.getTime()) ? null : date
}

File: server/internal/feature/finance/subscription/date.go
package subscription

import (
	"time"

	"github.com/kinqbert/finlo/server/internal/feature/finance/model"
)

func NextBillingDate(billingDay int, from time.Time) time.Time {
	location := from.Location()
	today := time.Date(from.Year(), from.Month(), from.Day(), 0, 0, 0, 0, location)
	next := billingDate(from.Year(), from.Month(), billingDay, location)
	if next.Before(today) {
		nextMonth := time.Date(from.Year(), from.Month()+1, 1, 0, 0, 0, 0, location)
		next = billingDate(nextMonth.Year(), nextMonth.Month(), billingDay, location)
	}
	return next
}

func ResolveNextPaymentDates(subscriptions []model.Subscription, from time.Time) {
	today := time.Date(from.Year(), from.Month(), from.Day(), 0, 0, 0, 0, from.Location())
	for index := range subscriptions {
		if !subscriptions[index].Active {
			continue
		}
		if subscriptions[index].NextPaymentDate != nil {
			persisted := subscriptions[index].NextPaymentDate.In(from.Location())
			persistedDate := time.Date(persisted.Year(), persisted.Month(), persisted.Day(), 0, 0, 0, 0, from.Location())
			if !persistedDate.Before(today) {
				continue
			}
		}
		next := NextBillingDate(subscriptions[index].BillingDay, from)
		subscriptions[index].NextPaymentDate = &next
	}
}

func billingDate(year int, month time.Month, preferredDay int, location *time.Location) time.Time {
	lastDay := time.Date(year, month+1, 0, 0, 0, 0, 0, location).Day()
	day := max(1, min(preferredDay, lastDay))
	return time.Date(year, month, day, 0, 0, 0, 0, location)
}
